web/hooks/use-persistent-state.ts

```diff
@@ -141,3 +146,81 @@ export const usePersistentState = <T>(
 
   return [state, setState] as const
 }
+
+export const usePersistentRevalidatedState = <T>(
+  initial: T,
+  persist: PersistenceOptions<T>,
+  revalidation: RevalidationOptions<T>
+) => {
+  const store = persist.store
+  // Is there any reason why a persistent store couldn't write a string?
+  const stringStore = store as PersistentStore<any>
+  const key = persist.key
+  const lastUpdateTimeKey = `${key}-last-update-time`
+  const [lastUpdateTime, setLastUpdateTime] = useState<number | undefined>(
+    undefined
+  )
+  const hasLoaded = useHasLoaded()
+  const { every, callback } = revalidation
+
+  // Note that it's important in some cases to get the state correct during the
+  // first render, or scroll restoration won't take into account the saved state.
+  // However, if this is the first server render, we don't want to read from the store,
+  // because it could cause a hydration error.
+  const savedValue =
+    hasLoaded && key != null && store != null ? store.get(key) : undefined
+
+  const [state, setState] = useStateCheckEquality(savedValue ?? initial)
+
+  useEffect(() => {
+    if (!hasLoaded || key === null || store === null) return
+
+    if (lastUpdateTime === undefined) {
+      const lastUpdateTimeTemp = parseInt(
+        stringStore.get(lastUpdateTimeKey) ?? 0
+      )
+      setLastUpdateTime(lastUpdateTimeTemp)
+      return
+    }
+    const now = Date.now()
+    if (lastUpdateTime + every < now) {
+      setLastUpdateTime(now)
+      stringStore.set(lastUpdateTimeKey, now.toString())
+      callback().then(setState)
+    }
+  }, [
+    hasLoaded,
+    store,
+    key,
+    lastUpdateTime,
+    every,
+    callback,
+    stringStore,
+    lastUpdateTimeKey,
+    setState,
+  ])
```

Comment:
This effect needs to be run after `every` time has elapsed. I'd remove the whole deps list so it runs on every re-render. Currently, it's mostly running just the first time you render a component with the hook, I think.

Reply:
Yeah the revalidation is kind of low stakes in this case imo so I didn't want to run it too often. But will probably be best if others use it that want more assurance of revalidation.